wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1194 @@
+    bytes_to_hex(&ins, out + 2, 1);
+    bytes_to_hex(&p1, out + 4, 1);
+    bytes_to_hex(&p2, out + 6, 1);
+    bytes_to_hex(&lc, out + 8, 1);
+    out[10] = 0;
+    return (int)XSTRLEN(cmd);
+}
+
+static int iotsafe_cmd_add_tlv_ex(char *cmd, byte tag, uint16_t len,
+        const byte *val, int taglen_size)
+{
+    word32 cur_csim_len;
+    word32 off;
+    byte cur_lc;
+    char *out;
+    int ret;
+    byte *len_b = (byte *)&len;
+    char *len_csim_str = (cmd + AT_CSIM_CMD_SIZE - 5);
+
+    if (XSTRLEN(cmd) < AT_CSIM_CMD_SIZE)
+        return BAD_FUNC_ARG;
+
+    if ((taglen_size < 1) || (taglen_size > 2))
+        return BAD_FUNC_ARG;
+
+    /* Read out current CSIM len from the existing string.
+     * The generated command may have the format:
+     *  "AT+CSIM= 68" (< 100: leading space)
+     *   or
+     *  "AT+CSIM=212" (>=100: same positions, leading hundreds)
+     *
+     */
+    cur_csim_len = (word32)len_csim_str[2] - '0';
+    cur_csim_len += (len_csim_str[1] - '0') * 10;
+    if (len_csim_str[0] >= '0' &&
+            len_csim_str[0] <= '9') {
+        cur_csim_len += len_csim_str[0] * 100;
+    } else if (len_csim_str[0] != ' ') {
+        return BAD_FUNC_ARG;
+    }
+
+    if (cur_csim_len > IOTSAFE_CMDSIZE_MAX) {
+        return MEMORY_E;
+    }
+
+    /* Read out current Lc parameter in the CSIM command, last byte in the
+     * header
+     */
+    if (hex_to_bytes(cmd + AT_CSIM_CMD_SIZE + AT_CMD_LC_POS, &cur_lc, 1) < 0)
+        return BAD_FUNC_ARG;
+
+    /* Increase Lc and CSIM length according to the TLV len */
+    cur_lc += 1 + taglen_size + len;
+    cur_csim_len += 2 + (2 * taglen_size) + 2*len;
+
+    /* Position the out buffer to the current end of string */
+    out = cmd + XSTRLEN(cmd);
+    off = 0;
+
+    /* Write Tag */
+    bytes_to_hex(&tag, out, 1);
+    off += 2;
+
+    /* Write Len */
+    if (taglen_size == 2) {
+        bytes_to_hex(&len_b[1], out + 2, 1);
+        bytes_to_hex(&len_b[0], out + 4, 1);
+        off += 4;
+    } else {
+        bytes_to_hex(&len_b[0], out + 2, 1);
+        off += 2;
+    }
+
+    /* Write Val */
+    ret = bytes_to_hex(val, out + off, len);
+    off += ret;
+
+    /* Terminate string */
+    out[off] = 0;
+
+    /* Write new CSIM command size */
+    if (cur_csim_len > 99)
+        len_csim_str[0] = (cur_csim_len / 100) + '0';
+    else
+        len_csim_str[0] = ' ';
+    len_csim_str[1] = ((cur_csim_len % 100) / 10) + '0';
+    len_csim_str[2] = (cur_csim_len % 10) + '0';
+
+    /* Write new Lc value */
+    bytes_to_hex(&cur_lc, cmd + AT_CSIM_CMD_SIZE + AT_CMD_LC_POS, 1);
+    return off;
+}
+
+static int iotsafe_cmd_add_tlv16(char *cmd, byte tag, uint16_t len,
+        const byte *val)
+{
+    return iotsafe_cmd_add_tlv_ex(cmd, tag, len, val, 2);
+}
+
+static int iotsafe_cmd_add_tlv(char *cmd, byte tag, byte len, const byte *val)
+{
+    return iotsafe_cmd_add_tlv_ex(cmd, tag, len, val, 1);
+}
+
+static void iotsafe_cmd_complete(char *cmd)
```

Comment:
Are you sure that this will never overflow? Maybe it would be good to pass in a length parameter to make sure that `XSTRLEN(cmd) + 4` doesn't go over the buffer? And if it would just put a `\0` at the end of the buffer and log an error? I guess this question goes for all of the `iotsafe_cmd_*` API.

Reply:
these functions are for static use only (within the module). It's implicit that the buffer should be as big as `IOTSAFE_CMDSIZE_MAX`.

Reply:
So the command may never be larger than `IOTSAFE_CMDSIZE_MAX`. Maybe having at least a check in these functions that we never access anything more than `IOTSAFE_CMDSIZE_MAX`?

Reply:
`iotsafe_cmd_add_tlv_ex`:

This is done further down before writing to the buffer (it would return MEMORY_E). However, it does not take into account the XSTRLEN part. Fixing.